packages/base-controller/src/BaseControllerV2.ts

```diff
@@ -245,34 +259,38 @@ export function getPersistentState<S extends Record<string, Json>>(
  * @param metadataProperty - The metadata property to use to derive state.
  * @returns The metadata-derived controller state.
  */
-function deriveStateFromMetadata<S extends Record<string, Json>>(
-  state: S,
-  metadata: StateMetadata<S>,
+function deriveStateFromMetadata<ControllerState extends Record<string, Json>>(
+  state: ControllerState,
+  metadata: StateMetadata<ControllerState>,
   metadataProperty: 'anonymous' | 'persist',
 ): Record<string, Json> {
-  return Object.keys(state).reduce((persistedState, key) => {
-    try {
-      const stateMetadata = metadata[key as keyof S];
-      if (!stateMetadata) {
-        throw new Error(`No metadata found for '${key}'`);
+  return Object.keys(state).reduce(
+    (
+      persistedState: Record<keyof ControllerState, Json>,
```

Comment:
Applied here: ee2467dafecd8ca83c736589ff9462e74421ee38.

This seems like a fundamental limitation of `reduce`, as it doesn't distinguish between the initial value type and the eventual return type. It might be worth writing a utility type for typing reduce that lets us specify both, with the initial value type defaulting to `Partial<ReturnValueType>`.

`key` resolving to `string` is probably because the return type of `Object.keys()` defaults to `string[]` without further inference-based narrowing. This casting gives us good intellisense/autocomplete: `Object.keys(state) as (keyof ControllerState)[]`.